### PR TITLE
WorkerCheck suppression, messages

### LINKED ISSUE #38: ErNamedClass.java:35-38: Annotations to skip rule. It is...
The puzzle `35-ba2ad0f0` from #35 has to be resolved: 

https://github.com/l3r8yJ/oop-cop/blob/b6cabb725bb75b5c8b634a8a7324423484005b2c/src/main/java/ru/l3r8y/rule/ErNamedClass.java#L35-L38

The puzzle was created by @rultor on 07-Apr-23. 

 role: DEV. 

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by h1alexbel**
@l3r8yJ take a look, please
closes #38

---

## PR-Codex overview
This PR focuses on adding suppression annotations to the codebase to suppress specific OOP rules. 

### Detailed summary
- Added `@SuppressWarnings` annotation to `User`, `SuppressedWorker`, `Marked`, `Mutable`, `Parser` classes.
- Updated `pom.xml` to change the value of `maximum` from 2 to 4.
- Updated `MethodChangesState.java` to include a link in the complaint message.
- Updated `ClassNames.java` to skip the rule check if the class is suppressed.
- Updated `ClassMethods.java` to skip the method check if the class is suppressed.
- Renamed `ErNamedClass` to `ErSuffixCheck`.
- Added `ErSuffixCheck` rule to check for classes ending with '-er' suffix.
- Added `IsSuppressed` class to check if a specific check is suppressed.
- Updated test cases for the changes made.

> The following files were skipped due to too many changes: `src/main/java/ru/l3r8y/parser/IsSuppressed.java`, `src/test/java/ru/l3r8y/parser/IsSuppressedTest.java`, `src/test/java/ru/l3r8y/parser/ClassNamesTest.java`, `src/main/java/ru/l3r8y/parser/SuppressedChecks.java`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by zoeself** [comment]
@l3r8yJ please review this Pull Request. Deadline (when it should be merged or closed) is ``2024-01-06T08:30:25.223586``.

You should check if the requirements have been implemented (partially or in full), if there are unit tests covering the changes and if the CI build passes. Feel free to reject the PR or ask for changes if it's too big or not clear enough.

Estimation here is ``30 minutes``, that's how much you will be paid. You will be paid even if this PR gets rejected.

**Review by l3r8yJ — CHANGES_REQUESTED** [review]
@h1alexbel i think we should move from our custom annotations to suppress convention
like `@SuppressWarning("OOP.ErNamedClass")`

**Comment by h1alexbel** [comment]
@l3r8yJ fully agree, but for now I don't know how to do it in a right way, any ideas?

**Comment by h1alexbel** [comment]
as far as I know, I can check that annotation `@SuppressWarnings` exists on top of class, but how to check its value (rule name)?

**Comment by l3r8yJ** [comment]
> @l3r8yJ fully agree, but for now I don't know how to do it in a right way, any ideas?

You can look at [jtcop](https://github.com/volodya-lombrozo/jtcop) implementation of this, I haven't seen it, but I assume there is some object that scans the project only for these exclusions and then during the validation phase just skips these exclusions
It's probably similar to a pattern strategy where `Map<Rule, List<Exclusion>>` has a list of class/method names to be skipped in `Map<Rule, List<Exclusion>>` against the rule

We can also ask @volodya-lombrozo directly how best to do this?

**Comment by h1alexbel** [comment]
@l3r8yJ I'm finishing my own implementation, very soon I will submit it

**Comment by l3r8yJ** [comment]
> as far as I know, I can check that annotation `@SuppressWarnings` exists on top of class, but how to check its value (rule name)?

I think only with `CGLIB` or `Reflections`,

- `CGLIB` if classes don't have interfaces
- `Reflections` if classes have interfaces

**Comment by h1alexbel** [comment]
@l3r8yJ take a look, please

File: src/test/java/ru/l3r8y/extensions/ManySuppressions.java [hunk 1]
```diff
@@ -21,23 +21,34 @@
...
+        final ParameterContext pctx,
+        final ExtensionContext ectx
+    ) {
+        return Objects.equals(pctx.getParameter().getType(), Path.class);
+    }
+
+    @Override
+    public Object resolveParameter(
+        final ParameterContext pctx,
+        final ExtensionContext ectx) {
```
Comment (l3r8yJ) [inline src/test/java/ru/l3r8y/extensions/ManySuppressions.java:51]:
```suggestion
        final ExtensionContext ectx
    ) {
```

File: pom.xml [hunk 2]
```diff
@@ -257,7 +257,7 @@
                     <limit>
                       <counter>CLASS</counter>
                       <value>MISSEDCOUNT</value>
-                      <maximum>2</maximum>
+                      <maximum>3</maximum>
```
Comment (l3r8yJ) [inline pom.xml:260]: what is this parameter responsible for?

File: src/it/correct/src/main/java/Parser.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import java.io.File;
+
+@SuppressWarnings({"OOP.WorkerCheck", "OOP.MutableStateCheck"})
```
Comment (l3r8yJ) [inline src/it/correct/src/main/java/Parser.java:3]: I'm not sure about the `Worker`, for programmer who has never seen this concept before, term `Worker` won't be obvious

File: src/main/java/ru/l3r8y/parser/ClassNames.java [hunk 4]
```diff
@@ -76,10 +75,19 @@ public Collection<ClassName> all() {
      */
     private void addIfEndsWithEr(final Node clazz) {
         if (clazz instanceof ClassOrInterfaceDeclaration) {
-            final String name = ClassOrInterfaceDeclaration.class
-                .cast(clazz)
-                .getNameAsString();
-            this.accum.add(new ParsedClassName(name, this.path));
+            final ClassOrInterfaceDeclaration declaration =
+                ClassOrInterfaceDeclaration.class.cast(clazz);
+            if (!new IsSuppressed(
```
Comment (l3r8yJ) [inline src/main/java/ru/l3r8y/parser/ClassNames.java:80]: formatting

File: src/main/java/ru/l3r8y/parser/IsSuppressed.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    ) {
+        this.declaration = dclrtn;
+        this.check = chk;
+    }
+
+    // @checkstyle ReturnCountCheck (40 lines).
+    @Override
+    @SuppressWarnings({"PMD.OnlyOneReturn", "PMD.AvoidDeeplyNestedIfStmts"})
+    public Boolean value() {
+        final Optional<AnnotationExpr> annotation =
```
Comment (l3r8yJ) [inline src/main/java/ru/l3r8y/parser/IsSuppressed.java:87]: too big [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity), i think you need to decompose this method somehow

File: src/main/java/ru/l3r8y/rule/WorkerCheck.java [hunk 6]
```diff
@@ -53,7 +48,12 @@ public final class ErNamedClass implements Rule {
     public Collection<Complaint> complaints() {
         return new ConditionRule(
             this::isEndsWithEr,
-            new WrongClassNaming(this.name, "class ends with '-er' suffix")
+            new WrongClassNaming(
+                this.name,
+                // @checkstyle StringLiteralsConcatenationCheck (4 lines).
+                "class ends with '-er' suffix, it's prohibited,"
+                + " read: https://www.yegor256.com/2015/03/09/objects-end-with-er.html"
```
Comment (l3r8yJ) [inline src/main/java/ru/l3r8y/rule/WorkerCheck.java:55]: not sure about `+` concatenation here

**Review by l3r8yJ — CHANGES_REQUESTED** [review]
@h1alexbel take a look at my comments below, please

File: src/main/java/ru/l3r8y/rule/WorkerCheck.java (hunk 6 above)
Comment (l3r8yJ) [reply src/main/java/ru/l3r8y/rule/WorkerCheck.java:55]: i think you can add some decorator for `Rule` like `LinkedRule` with link to resource, which will be added to warning message

File: src/main/java/ru/l3r8y/parser/IsSuppressed.java (hunk 5 above)
Comment (h1alexbel) [reply src/main/java/ru/l3r8y/parser/IsSuppressed.java]: @l3r8yJ check latest commit and puzzles in class header

File: src/main/java/ru/l3r8y/rule/WorkerCheck.java (hunk 6 above)
Comment (l3r8yJ) [reply src/main/java/ru/l3r8y/rule/WorkerCheck.java:55]: go ahead

File: src/it/correct/src/main/java/Parser.java (hunk 3 above)
Comment (h1alexbel) [reply src/it/correct/src/main/java/Parser.java:3]: @l3r8yJ how to name it then?
`ErNameCheck`?

File: pom.xml (hunk 2 above)
Comment (h1alexbel) [reply pom.xml]: classes without tests

File: src/main/java/ru/l3r8y/parser/ClassNames.java (hunk 4 above)
Comment (h1alexbel) [reply src/main/java/ru/l3r8y/parser/ClassNames.java]: fixed

File: src/it/correct/src/main/java/Parser.java (hunk 3 above)
Comment (l3r8yJ) [reply src/it/correct/src/main/java/Parser.java:3]: maybe `ErSuffixCheck`

File: src/test/java/ru/l3r8y/extensions/ManySuppressions.java (hunk 1 above)
Comment (h1alexbel) [reply src/test/java/ru/l3r8y/extensions/ManySuppressions.java]: fixed

**Comment by h1alexbel** [comment]
@l3r8yJ take a look, please

**Review by l3r8yJ — APPROVED** [review]
@h1alexbel looks good to me!

**Comment by l3r8yJ** [comment]
@rultor merge

**Comment by zoeself** [comment]
@l3r8yJ thank you for resolving this ticket. I've just added it to your active invoice. You can always check all your invoices and more on the [Contributor Dashboard](https://self-xdsd.com).

**Pull request merged**